### PR TITLE
automated release research

### PR DESCRIPTION
**Pull request by Kargathia**
Description of research done for automation of release pipeline.

Will eventually be part of Fontys internship report, but should also be useful for contributors wishing to get background information on design decisions.

### CODE REVIEW

File: research/automated_release.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+The default Git Flow model does not describe forked repositories making upstream pull requests. The forked model is attractive to projects with open-source contributors, as it allows contributions without granting write access to strangers.
+
+Contributors fork the base repository on GitHub into their own. At that point they can create their own branches, or use their local version of `develop` - the upstream repository doesn't care. When they are done, they make a pull request to the correct target branch on the base repository. Pull requests can be commented on by many, but only merged by somebody with write access to the base repository.
+
+We'll refer to the base repository as `upstream`, and to the contributor repository as `downstream`.
+
+In a forked Git Flow repository, there will be no `feature` or `hotfix` branches in `upstream`. This is all done in `downstream`.
+
+Contributors fork new `downstream/feature` branches from `upstream/develop`, and when ready, make a pull request back into `upstream/develop`.
```
Comment (elcojacobs) [inline research/automated_release.md]: forking -> branching

File: research/automated_release.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+`release` branches are created in `upstream`. If any fixes need to be made, these are directly done in `downstream/release` (or a forked branch), and merged back in through pull request.
+
+The same goes for `hotfix` branches: `downstream/hotfix` is forked from `upstream/master`, and merged back into `upstream/master`.
+
+At any given time, there are at most three active branches in `upstream`: `master`, `develop`, and (when active) `release`.
```
Comment (elcojacobs) [inline research/automated_release.md]: In general, only 3 branches will exist. Exceptions are:
- Long running feature branches that involve multiple developers. They can use a feature branch on the upstream repo to synchronize work. They could do PRs to the feature branch.
- Multiple release branches can exist in parallel, for example the latest stable release and a new release candidate for the next major version.

In general: small features do not require a feature branches on the upstream repo.

File: research/automated_release.md [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+>
+> Additional labels for pre-release and build metadata are available as extensions to the MAJOR.MINOR.PATCH format.
+
+This can be used without any adjustments or special cases by the Brewblox project.
+
+## Tagging
+
+Git itself supports [tags][12]. These can provide extra information about the repository. Specifically, they can be used to mark where releases were done, so that the HEAD can easily be set at that point.
+
+GitHub offers additional functionality for tags with [GitHub Releases][13].
```
Comment (elcojacobs) [inline research/automated_release.md]: github releases as changelog

File: research/automated_release.md (hunk 1 above)
Comment (Kargathia) [reply research/automated_release.md]: fixed everywhere: branch creation is now referred to as branching.

File: research/automated_release.md (hunk 2 above)
Comment (Kargathia) [reply research/automated_release.md]: Fixed

File: research/automated_release.md (hunk 3 above)
Comment (Kargathia) [reply research/automated_release.md]: Fixed

**Pull request merged**